Simplenote/CoreDataManager.swift

```diff
@@ -0,0 +1,87 @@
+//
+//  CoreDataManager.swift
+//  Simplenote
+//
+//  Created by Charlie Scheer on 5/23/24.
+//  Copyright © 2024 Simperium. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+enum CoreDataManagerError: Error {
+    case couldNotBuildModel
+    case noApplicationFilesDirectoryURL
+    case foundNotDirectoryAtFilesDirectoryURL
+
+    var description: String {
+        switch self {
+        case .couldNotBuildModel:
+            return "Could not build model from URL"
+        case .noApplicationFilesDirectoryURL:
+            return "No url found for the application files directory"
+        case .foundNotDirectoryAtFilesDirectoryURL:
+            return "Item at applications files url is not a directory"
+        }
+    }
+}
+
+@objcMembers
+class CoreDataManager: NSObject {
+    private(set) var managedObjectModel: NSManagedObjectModel
+    private(set) var managedObjectContext: NSManagedObjectContext
+    private(set) var persistentStoreCoordinator: NSPersistentStoreCoordinator
+
+    init(storageSettings: StorageSettings = StorageSettings()) throws {
+        guard let modelURL = storageSettings.modelURL,
+              let mom = NSManagedObjectModel(contentsOf: modelURL) else {
+            throw CoreDataManagerError.couldNotBuildModel
+        }
+
+        let context = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
+        let psc = try Self.preparePSC(with: storageSettings, model: mom)
+
+        self.managedObjectModel = mom
+        self.managedObjectContext = context
+        self.persistentStoreCoordinator = psc
+    }
+
+    static private func preparePSC(with storageSettings: StorageSettings, model: NSManagedObjectModel) throws -> NSPersistentStoreCoordinator {
+        guard let applicationFilesDirectory = storageSettings.applicationFilesDirectory else {
+            throw CoreDataManagerError.noApplicationFilesDirectoryURL
+        }
+
+        // Validate the directory for the store DB
+        do {
+            try Self.validateResourceValueForDirectory(at: applicationFilesDirectory)
+        } catch {
+            try handleDirectoryError((error as NSError), directoryURL: applicationFilesDirectory)
+        }
```

Comment:
IMHO L50-L59 should probably live within a single method?

You'd just encapsulate all of the URL logic in a single place, and the "preparePSC" method would probably end up being super compact

```
func persistentStoreURL() throws -> URL {

}
``